hazelcast/src/main/java/com/hazelcast/internal/affinity/ThreadAffinitySupport.java

```diff
@@ -0,0 +1,164 @@
+
+    private ThreadAffinitySupport(ILogger logger) {
+        this.logger = logger;
+    }
+
+    public ThreadAffinityController create(Class<? extends Thread> cls) {
+        checkNotNull(cls);
+        ThreadAffinity type = cls.getAnnotation(ThreadAffinity.class);
+        if (type != null) {
+            return getOrPutIfAbsent(registry, type, key -> new ThreadAffinityController(logger, type));
+        }
+
+        return null;
+    }
+
+    public static void init(ILogger logger) {
+        if (!isAffinityEnabled()) {
+            return;
+        }
+
+        if (!INIT.compareAndSet(false, true)) {
+            throw new IllegalStateException("Thread affinity support disallows multiple Hazelcast instances under the same JVM");
+        }
+
+        if (!isUnixFamily()) {
+            throw new IllegalStateException("Thread affinity not supported on this platform: " + OS);
+        }
+
+        logger.info("Thread affinity option detected, checking for dependencies");
+
+        try {
+            // Check if dependency OpenHFT Affinity is present in classpath
+            Class.forName("net.openhft.affinity.AffinityLock");
+        } catch (ClassNotFoundException e) {
+            logger.warning("Dependency net.openhft:affinity not found in classpath. Affinity support is disabled.");
+            return;
+        }
+
+        try {
+            // Check if JNA is available in the classpath
+            Class.forName("com.sun.jna.Platform");
+        } catch (ClassNotFoundException e) {
+            logger.warning("Dependency net.java.dev.jna:jna* not found in classpath. Affinity support is disabled.");
+        }
+
+        if (!disjoint(ThreadAffinityProperties.getCoreIds(IO), ThreadAffinityProperties.getCoreIds(PARTITION_THREAD))) {
+            logger.warning("Affinity assignments for the different affinity groups have some cores in common (shared).");
+        }
+
+        failFastChecks();
+
+        logger.info("Thread affinity dependencies available, support enabled");
+        INSTANCE = new ThreadAffinitySupport(logger);
+    }
+
+    private static void failFastChecks() {
+        // Available resources conflicts
+        int totalAvailableCores = Runtime.getRuntime().availableProcessors();
+        int partitionOpCores = ThreadAffinityProperties.countCores(PARTITION_THREAD);
+        if (partitionOpCores > totalAvailableCores) {
+            throw new IllegalStateException("Thread affinity core count for " + PARTITION_THREAD + " is set to high, "
+                    + "more than the available core count on the system " + totalAvailableCores);
+        }
+
+        int ioCores = ThreadAffinityProperties.countCores(IO);
+        if (ioCores > totalAvailableCores) {
+            throw new IllegalStateException("Thread affinity core count for " + IO + " is set to high, "
+                    + "more than the available core count on the system " + totalAvailableCores);
+        }
+
+        if (ioCores % 2 != 0) {
```

Comment:
why does this need to be even?

Reply:
It overrides the IO thread count, so its split among input & output threads.